Add IncomingMessageProperties injection support post (#18)
CoreWCF/corewcf.github.io: _posts/2024-05-06-message-property-injection.markdown

---
# Layout
layout: post
title:  "Add IncomingMessageProperty injection support anf Kafka transport enhancements in CoreWCF"

> date:   2024-05-06 13:00:00 +0200

Reviewer: Update the date to something more current.

Author: Updated!

categories: release
# Author
author: Guillaume Delahaye (https://github.com/g7ed6e)
---
### Introduction

> The CoreWCF 1.6.0 release introduced a new feature that allows to inject incoming message properties as operation contract parameter.

Reviewer: ```suggestion
The CoreWCF 1.6.0 release introduced a new feature that enables injecting incoming message properties as operation contract parameters.
```

### Implementation

> This feature is implemented in the `OperationContractParameterGenerator` which already support injecting service registered in the DI container or `HttpContext` into the operation contract parameters.

Reviewer: ```suggestion
This feature is implemented in the `OperationContractParameterGenerator` which already support injecting services registered in the DI container or `HttpContext` into the operation contract parameters.
```

The `InjectedAttribute.PropertyName` is exposed to specify the message property to be injected.
```c#
namespace CoreWCF
{
    [AttributeUsage(AttributeTargets.Parameter)]
    public sealed class InjectedAttribute : Attribute
    {
        public string PropertyName { get; set; }
    }
}
```
### NetTcpBinding usage

> A service exposing a `NetTcpBinding` can pull the `RemoteEndpointMessageProperty`.

Reviewer: ```suggestion
A service exposing a `NetTcpBinding` can inject the `RemoteEndpointMessageProperty`.
```

```c#
public interface IHelloWorldService
{
    string SayHello();
}

public class HelloWorldService : IHelloService
{
    public string SayHello([Injected(PropertyName = RemoteEndpointMessageProperty.Name)] RemoteEndpointMessageProperty remoteEndpointMessageProperty)
    {
        return $"Hello from {remoteEndpointMessageProperty.Address}:{remoteEndpointMessageProperty.Port}";
    }
}
```
### BasicHttpBinding usage

> A service exposing a `BasicHttpBinding` can pull the `HttpRequestMessageProperty`.

Reviewer: ```suggestion
A service exposing a `BasicHttpBinding` can inject the `HttpRequestMessageProperty`.
```

    {
        foreach (IHeader messageHeader in consumeResult.Message.Headers)
        {
            _headers.Add(new KafkaMessageHeader(messageHeader.Key, messageHeader.GetValueBytes()));
        }

        PartitionKey = consumeResult.Message.Key;
        Topic = consumeResult.Topic;
    }
    
    public IReadOnlyCollection<KafkaMessageHeader> Headers => _headers as IReadOnlyCollection<KafkaMessageHeader>;
    public ReadOnlyMemory<byte> PartitionKey { get; }
    public string Topic { get; }
}
```

Client side the partition key and the headers can be provided using an `OperationContextScope`.
```c#
using (var scope = new System.ServiceModel.OperationContextScope((System.ServiceModel.IContextChannel)channel))
{
    ServiceModel.Channels.KafkaMessageProperty outgoingProperty = new();
    outgoingProperty.Headers.Add(new ServiceModel.Channels.KafkaMessageHeader("header1", Encoding.UTF8.GetBytes("header1Value")));
    outgoingProperty.PartitionKey = Encoding.UTF8.GetBytes("key");
    System.ServiceModel.OperationContext.Current.OutgoingMessageProperties[ServiceModel.Channels.KafkaMessageProperty.Name] =
        outgoingProperty;
    channel.DoSomething();
}
```

> Service side the implementer can get these values back by pulling the `KafkaLessageProperty`

Reviewer: ```suggestion
Service side the implementer can get these values back by injecting the `KafkaMessageProperty`
```
